controllers/istio_extension_reconciler.go

```diff
@@ -0,0 +1,258 @@
+	}
+
+	logger.V(1).Info("building wasm configs for istio extension", "effectivePolicies", len(effectivePolicies.(EffectiveRateLimitPolicies)))
+
+	wasmActionSets := kuadrantgatewayapi.GrouppedHTTPRouteMatchConfigs{}
+
+	// build the wasm policies for each topological path that contains an effective rate limit policy affecting an istio gateway
+	for pathID, effectivePolicy := range effectivePolicies.(EffectiveRateLimitPolicies) {
+		gatewayClass, gateway, _, _, _, _ := common.ObjectsInRequestPath(effectivePolicy.Path)
+
+		// ignore if not an istio gateway
+		if gatewayClass.Spec.ControllerName != istioGatewayControllerName {
+			continue
+		}
+
+		wasmActionSetsForPath, err := wasm.BuildActionSetsForPath(pathID, effectivePolicy.Path, effectivePolicy.Spec.Rules(), rateLimitWasmActionBuilder(pathID, effectivePolicy, state))
+		if err != nil {
+			logger.Error(err, "failed to build wasm policies for path", "pathID", pathID)
+			continue
+		}
+		wasmActionSets.Add(gateway.GetLocator(), wasmActionSetsForPath...)
+	}
+
+	wasmConfigs := lo.MapValues(wasmActionSets.Sorted(), func(configs kuadrantgatewayapi.SortableHTTPRouteMatchConfigs, _ string) wasm.Config {
+		return wasm.BuildConfigForActionSet(lo.Map(configs, func(c kuadrantgatewayapi.HTTPRouteMatchConfig, _ int) wasm.ActionSet {
+			return c.Config.(wasm.ActionSet)
+		}))
+	})
+
+	return wasmConfigs, nil
+}
+
+// buildIstioWasmPluginForGateway builds a desired WasmPlugin custom resource for a given gateway and corresponding wasm config
+func buildIstioWasmPluginForGateway(gateway *machinery.Gateway, wasmConfig wasm.Config) *istioclientgoextensionv1alpha1.WasmPlugin {
+	wasmPlugin := &istioclientgoextensionv1alpha1.WasmPlugin{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       kuadrantistio.WasmPluginGroupKind.Kind,
+			APIVersion: istioclientgoextensionv1alpha1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      wasm.ExtensionName(gateway.GetName()),
+			Namespace: gateway.GetNamespace(),
+			Labels:    KuadrantManagedObjectLabels(),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         gateway.GroupVersionKind().GroupVersion().String(),
+					Kind:               gateway.GroupVersionKind().Kind,
+					Name:               gateway.Name,
+					UID:                gateway.UID,
+					BlockOwnerDeletion: ptr.To(true),
+					Controller:         ptr.To(true),
+				},
+			},
+		},
+		Spec: istioextensionsv1alpha1.WasmPlugin{
+			TargetRefs: []*istiov1beta1.PolicyTargetReference{
+				{
+					Group: machinery.GatewayGroupKind.Group,
+					Kind:  machinery.GatewayGroupKind.Kind,
+					Name:  gateway.GetName(),
+				},
+			},
+			Url:          WASMFilterImageURL,
+			PluginConfig: nil,
+			Phase:        istioextensionsv1alpha1.PluginPhase_STATS, // insert the plugin before Istio stats filters and after Istio authorization filters.
+		},
+	}
+
+	if len(wasmConfig.ActionSets) == 0 {
+		utils.TagObjectToDelete(wasmPlugin)
```

Comment:
I have a funny feeling about this. It seems like we should have know that we are meant to delete object before we ever got this far. To me once we created the the `wasmConfig` we should know that all the resources for that config should be deleted. 

Again this is not something I am looking to change. Just making a note of it.